### PR TITLE
Added findURLs, isValid, toTitleCase functions and their corresponding test files

### PR DESCRIPTION
**Pull request by leishab**

### CODE REVIEW

**Comment by gabeabrams** [comment]
In `/src/index.ts`, import your helper functions in the section where all the helpers are imported, then export your helper functions below in the section marked by `// Helpers`

**Comment by leishab** [comment]
**findURLs:**
This function iterates through and identifies URLs in a block of text, after which it returns their start and end index. The function contains a while loop to use the regex to locate and then match the URL(s) within the block of text, after which the start and end index of the URL is found and pushed to the found array, which is returned to the user. 

The test file for this function involves various test cases which test whether the function returns the correct start and end indices for the correctly identified URLs, edge cases such as when the URL has a trailing or leading punctuation mark and returns empty array for strings without valid URLs or for invalid input formats.

**toTitleCase:**
This function changes any input string so that it adheres to the rules of Chicago Title Case Style. Before this function can proceed to do this, it first makes sure to throw an error if the input is undefined, null or not a string. It then converts the entire input string to lowercase and splits it into separate words. It maps over this string and capitalizes the first and last word, and any word in the input that is not a part of the lowercaseWords list. It then puts the words back together and returns the new string that has been converted to Chicago Title Case Style.

It also includes the test file to validate the function's behavior. It first tests the function with some tests that would return valid, some edge cases such as single letters, all capitalized strings etc. and tests whether the appropriate error is thrown incase an input is invalid
 
 **isValid**:
This function checks if a given input string is a valid URL. It does so by using the input URL to create a URL object and using the URL constructor to throw an error if the URL is invalid. 

It also includes a test file to verify the function's correctness. The test file consists of many examples of URLs that would be considered valid, as well as invalid.

**Review by gabeabrams — COMMENTED** [review]
More comments

File: src/helpers/validators/isValid.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * @param URL The input string that needs checked as a URL or not.
+ * @returns A true boolean value if the input string is a valid URL, and a false
+ *          boolean value if the input string is a invalid URL
+ */
+const isValid = (url: string): boolean => {
+  try {
+    const parsed = new URL(url);
+
+    // Checks that the URL's protocol is valid
+    if (!['https:', 'http:', 'file:', 'ftp:'].includes(parsed.protocol)) {
```
Comment (gabeabrams) [inline src/helpers/validators/isValid.ts:13]: Only https and http

File: src/helpers/validators/isValid.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      return false;
+    }
+
+    // Checks that the hostname is not empty and doesn't start with a dot
+    if (!parsed.hostname || parsed.hostname.startsWith('.')) {
```
Comment (gabeabrams) [inline src/helpers/validators/isValid.ts:18]: `parsed.hostname.trim().length === 0

File: src/helpers/validators/isValid.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      return false;
+    }
+
+    // Checks that the hostname is not empty and doesn't end with a dot
+    if (!parsed.hostname || parsed.hostname.endsWith('.')) {
```
Comment (gabeabrams) [inline src/helpers/validators/isValid.ts:23]: Combine with above

File: src/helpers/validators/isValid.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      return false;
+    }
+
+    // Checks that the URL doesn't have consecutive periods in the hostname
+    if (/(\.\.)/.test(parsed.hostname)) {
```
Comment (gabeabrams) [inline src/helpers/validators/isValid.ts:28]: parsed.hostname.includes('..')

File: src/helpers/validators/isValid.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      return false;
+    }
+
+    // Checks that the hostname is not just numbers
+    if (!parsed.hostname || parsed.hostname.endsWith('.') || /^\d+$/.test(parsed.hostname)) {
```
Comment (gabeabrams) [inline src/helpers/validators/isValid.ts:33]: This doesn't seem necessary

File: src/helpers/validators/isValid.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      return false;
+    }
+
+    // Checks that if the port is given, it is a valid number
+    if (parsed.port && !/^\d+$/.test(parsed.port)) {
```
Comment (gabeabrams) [inline src/helpers/validators/isValid.ts:38]: `Number.isNaN(Number.parseInt(parsed.port, 10))`

File: src/helpers/validators/toTitleCase.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+/**
+ * Converts a string to title case based on Chicago Title Case Style rules.
+ * @author Leisha Bhandari
+ * @param input The string that needs to be converted to Chicago title case.
+ * @returns Input string converted to title case
+ */
+
+const toTitleCase = (input: any): string => {
```
Comment (gabeabrams) [inline src/helpers/validators/toTitleCase.ts:8]: Input should be a string

File: src/helpers/validators/toTitleCase.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  // Check if there input is undefined, if so, throws an error
+  if (input === undefined) throw new Error('Input is undefined');
+  // Check if there input is null, if so, throws an error
+  if (input === null) throw new Error('Input is null');
+  // Check if there input is a string, if not, throw an error
+  if (typeof input !== 'string') throw new Error('Input is a number. It must be a string');
```
Comment (gabeabrams) [inline src/helpers/validators/toTitleCase.ts:14]: We don't need these checks

File: src/helpers/validators/toTitleCase.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+  // List of lowercase words that must not be capitalized unless they're the
+  // first or last word of the input string
+  const lowerCaseWords: string[] = [
+    'in', 'nor', 'of', 'on', 'or', 'as', 'at', 'but', 'by',
+    'for', 'if', 'so', 'the', 'a', 'an', 'and', 'to', 'up',
+    'yet',
+  ];
```
Comment (gabeabrams) [inline src/helpers/validators/toTitleCase.ts:22]: These should exist outside the function

File: src/helpers/validators/toTitleCase.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    // Splits string into separate words
+    .split(' ')
+    .map((word, index, words) => {
+      // Capitalize the first, last word, and any word in the input that is not a part of the lowercaseWords list
+      return (index === 0 || index === words.length - 1 || !lowerCaseWords.includes(word)
+        ? `${word.charAt(0).toUpperCase()}${word.slice(1)}`
+        : word);
+    })
+    // Join the words back
+    .join(' ');
```
Comment (gabeabrams) [inline src/helpers/validators/toTitleCase.ts:36]: Formatting

```ts

return (
  input
    // The entire input string is converted to lowercase
    .toLowerCase()
    // Splits string into separate words
    .split(' ')
    .map((word, index, words) => {
      // Capitalize the first, last word, and any word in the input that is not a part of the lowercaseWords list
      return (index === 0 || index === words.length - 1 || !lowerCaseWords.includes(word)
        ? `${word.charAt(0).toUpperCase()}${word.slice(1)}`
        : word);
    })
    // Join the words back
    .join(' ')
);
```

File: src/helpers/validators/toTitleCase.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  return input
+  // The entire input string is converted to lowercase
+    .toLowerCase()
+    // Splits string into separate words
+    .split(' ')
+    .map((word, index, words) => {
+      // Capitalize the first, last word, and any word in the input that is not a part of the lowercaseWords list
+      return (index === 0 || index === words.length - 1 || !lowerCaseWords.includes(word)
+        ? `${word.charAt(0).toUpperCase()}${word.slice(1)}`
+        : word);
```
Comment (gabeabrams) [inline src/helpers/validators/toTitleCase.ts:33]: One item per line

```ts
return (
  (index === 0 || index === words.length - 1 || !lowerCaseWords.includes(word))
    ? `${word.charAt(0).toUpperCase()}${word.slice(1)}`
    : word
);
```

**Review by gabeabrams — CHANGES_REQUESTED** [review]
See comments. Super close!

File: src/helpers/validators/toTitleCase.test.ts [hunk 12]
```diff
@@ -84,6 +84,19 @@ test(
...
+    throw new Error('Input is undefined');
+  }
+  if (input === null) {
+    throw new Error('Input is null');
+  }
+  if (typeof input !== 'string') {
+    throw new Error('Input is a number. It must be a string');
+  }
+  return input;
+};
```
Comment (gabeabrams) [inline src/helpers/validators/toTitleCase.test.ts:98]: I don't understand this
